numpy/core/src/multiarray/arrayobject.c

```diff
@@ -1464,8 +1450,102 @@ array_richcompare(PyArrayObject *self, PyObject *other, int cmp_op)
                 n_ops.greater_equal);
         break;
     default:
-        result = Py_NotImplemented;
-        Py_INCREF(result);
+        Py_INCREF(Py_NotImplemented);
+        return Py_NotImplemented;
+    }
+    if (result == NULL) {
+        /*
+         * If the comparison resulted in NULL, then it failed with treating
+         * "other" as an array, either because it could not be converted at
+         * all, or because it had a flexible dtype (while self has not; this
+         * is checked earlier).  Our ideal behaviour would be:
+         * 1. For EQ and NE, if other cannot be converted or self and "other"
+         *    are scalars, return NotImplemented, so that python can assign
+         *    True of False as appropriate.
+         * 2. For EQ and NE, otherwise return a correctly shaped array of bool;
+         * 3. For LT, LE, GE, GT, return NotImplemented, so python can raise
+         *    a TypeError.
+         *
+         * However, for backwards compatibilty, we cannot yet return arrays,
+         * so we raise warnings instead for cases 2 and 3.
+         */
+        PyArrayObject *other_array = NULL;
+        PyArray_Descr *other_dtype = NULL;
+        int other_ndim = 0;
+        npy_intp other_dims[NPY_MAXDIMS];
+        int other_type_num, other_is_flexible;
+        /*
+         * Determine whether other has a flexible dtype; here, inconvertible
+         * is counted as inflexble.  (This repeats work done in the ufunc,
+         * but OK to waste some time in an unlikely path.)
+         */
+        if (PyArray_GetArrayParamsFromObject(
```

Comment:
I tried using the seemingly more logical `PyArray_FromAny(other, NULL, 0, 0, 0, NULL)` but somehow this returns a failure when other is a python string.

Reply:
You probably need to `PyErr_Fetch` above, before you start calling any more functions - most python functions are not willing to run if an exception is actively being thrown

Reply:
Ah, yes, that indeed does it!